core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSourceOutputSelector.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.exchange;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static io.trino.spi.exchange.ExchangeSourceOutputSelector.Selection.EXCLUDED;
+import static io.trino.spi.exchange.ExchangeSourceOutputSelector.Selection.INCLUDED;
+import static io.trino.spi.exchange.ExchangeSourceOutputSelector.Selection.UNKNOWN;
+import static java.lang.Math.max;
+import static java.util.Arrays.fill;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toMap;
+
+public class ExchangeSourceOutputSelector
+{
+    private static final long INSTANCE_SIZE = ClassLayout.parseClass(ExchangeSourceOutputSelector.class).instanceSize();
+
+    private final int version;
+    private final Map<ExchangeId, Slice> values;
+    private final boolean finalSelector;
+
+    // visible for Jackson
+    @JsonCreator
+    public ExchangeSourceOutputSelector(
+            @JsonProperty("version") int version,
+            @JsonProperty("values") Map<ExchangeId, Slice> values,
+            @JsonProperty("finalSelector") boolean finalSelector)
+    {
+        this.version = version;
+        this.values = Map.copyOf(requireNonNull(values, "values is null"));
+        this.finalSelector = finalSelector;
+    }
+
+    @JsonProperty
+    public int getVersion()
+    {
+        return version;
+    }
+
+    // visible for Jackson
+    @JsonProperty
+    public Map<ExchangeId, Slice> getValues()
+    {
+        return values;
+    }
+
+    @JsonProperty("finalSelector")
+    public boolean isFinal()
+    {
+        return finalSelector;
+    }
+
+    public Selection getSelection(ExchangeId exchangeId, int taskPartitionId, int attemptId)
```

Comment:
Is there a contract that we limit task to have at most 127 attempts. Feels reasonable but it should probably be enforced somewhere. In `TaskId` constructor?

Reply:
Yeah, that makes sense. Let's address it as a separate PR: https://github.com/trinodb/trino/issues/14459